src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs

```diff
@@ -38,7 +48,132 @@ public static IEnumerable<ApplicationPart> DiscoverAssemblyParts(string entryPoi
+                    {
+                        var fileName = Path.GetFileName(metadataAttribute.Value.Substring(0, metadataAttribute.Value.IndexOf(",")));
+                        var filePath = Path.Combine(Path.GetDirectoryName(assembly.Location), fileName);
+                        var additionalReference = new AdditionalReference
+                        {
+                            File = filePath,
+                            IncludeByDefault = metadataAttribute.Value.Substring(metadataAttribute.Value.IndexOf(",") + 1).Equals("true")
+                        };
+
+                        additionalAssemblies.Add(additionalReference);
+                        if (assembly.Equals(entryAssembly))
+                        {
+                            entryAssemblyAdditionalReferences.Add(additionalReference);
+                        }
+                    }
+                }
+                else
+                {
+                    // Fallback to loading the views like in previous versions if the additional reference metadata attribute is not present.
+                    var viewsAssemblies = GetViewAssemblies(assembly);
+                    foreach (var viewsAssembly in viewsAssemblies)
+                    {
+                        if (viewsAssembly != null)
+                        {
+                            var additionalReference = new AdditionalReference
+                            {
+                                File = null,
+                                Assembly = viewsAssembly,
+                                IncludeByDefault = true
+                            };
+
+                            additionalAssemblies.Add(additionalReference);
+                            if (entryAssembly.Equals(assembly))
+                            {
+                                entryAssemblyAdditionalReferences.Add(additionalReference);
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Load all the additional references.
+            for (var i = 0; i < additionalAssemblies.Count; i++)
+            {
+                var reference = additionalAssemblies[i];
+                if (reference.Assembly == null)
+                {
+                    var assembly = Assembly.LoadFile(reference.File);
+                    reference.Assembly = assembly;
+                }
+            }
+
+            // Discard any additional reference that was added by default through the dependency context.
+            var result = candidateAssemblies
+                .Where(ca => !additionalAssemblies.Any(ara => ara.Assembly.FullName.Equals(ca.FullName)))
+                .Distinct()
+                .OrderBy(assembly => entryAssembly.Equals(assembly) ? 0 : 1)
+                .ThenBy(a=> a.GetName().Name)
+                .ToList();
+
+            var entryAssemblyReferences = entryAssemblyAdditionalReferences.Select(e => e.Assembly).ToList();
```

Comment:
It would be nice to see a comment somewhere with a list of what the categories are, and what order you expect them to be in.